# Python Tutorial 1 – The Basics
**Print · Comments · Strings · Numbers · Data Structures · Booleans · Conditions**


**Plan for today**
1. Working with notebooks
2. `print` and comments
3. Strings and f-strings
4. Numbers, conversion and math
5. Lists, tuples, sets and dictionaries
6. Booleans and comparisons
7. Conditions (`if` / `elif` / `else`)
8. Common mistakes and best practices
9. Exercises


## 2. Print

`print` shows something on the screen. Text must be in quotes (`"` or `'`).
Click into the cell below and press <kbd>Shift</kbd>+<kbd>Enter</kbd>.

In [1]:
print("Hello beautiful people")

Hello beautiful people


By default, `print` adds a **newline** (`\n`) at the end, so the next print starts on a new line:

In [2]:
print("Hello")
print("Goodbye")

Hello
Goodbye


You can change that with the `end` argument, and print several things at once separated by commas:

In [3]:
print("Hello", end=" ")
print("Goodbye")

print("Python", "is", "fun", 2026)

Hello Goodbye
Python is fun 2026


You can also print the content of **variables**:

In [4]:
saying_hello = "Hello there!"
saying_goodbye = "Hasta la vista!"

print(saying_hello)
print(saying_goodbye)

Hello there!
Hasta la vista!


In [5]:
saying_hello

'Hello there!'

What happened here?
**In a notebook, the value of the last line in a cell is shown automatically.**
This is *not* normal Python behaviour: in a normal `.py` script you need `print` to see anything.

## 3. Comments

Python **ignores** everything after a `#`. Comments are useful to
- explain your code to other people (and to future you)
- temporarily switch off code without deleting it

In [ ]:
# This is a comment on its own line
'''
    This is a multi-line comment
    that spans multiple lines
'''
print("Commenting is great!")  # a comment next to code

Commenting is great!


Try it: comment out the first line and uncomment the second one.

In [7]:
print("Hello")
# print("Goodbye")

Hello


Shortcut to (un)comment many lines at once: select them and press <kbd>Ctrl</kbd>+<kbd>/</kbd>
(on some keyboard layouts <kbd>Ctrl</kbd>+<kbd>#</kbd>). Try it here:

In [ ]:
print("So")
print("many")
print("lines")
print("to")
print("(un-)")
print("comment")

So
many
lines
to
(un-)
comment


### Good comments
- The code already says **what** happens. Use comments to explain **why**.
- If you tried other approaches that didn't work, write down why. Future you will thank you.
- Put one space after the `#`: `# like this`, not `#like this`.
- More style tips: [PEP 8 – comments](https://peps.python.org/pep-0008/#comments)

## 4. Strings

Strings hold **text**. Create them with double quotes `"..."` or single quotes `'...'` – just don't mix them in one string.
A string is a **sequence of characters**.

| Method | What it does |
|---|---|
| `len(s)` | number of characters |
| `s[4]` | character at position 4 (counting starts at **0**!) |
| `sep.join(items)` | joins items into one string, with `sep` in between |
| `s.split(sep)` | splits the string at `sep` into a list |
| `s.find(sub)` | position of `sub` in `s` (or `-1` if not found) |
| `s.replace(old, new)` | replaces `old` with `new` |
| `s.upper()` / `s.lower()` | UPPER / lower case |
| `s.strip()` | removes spaces at the start and end |

More methods: [Python docs – string methods](https://docs.python.org/3/library/stdtypes.html#string-methods)

In [9]:
string_1 = "Hello"
string_2 = 'World'

type(string_1)

str

Math on strings? Yes! `+` **joins** (concatenates) strings:

In [10]:
string_1 + " " + string_2

'Hello World'

`*` **repeats** a string:

In [11]:
string_1 * 3

'HelloHelloHello'

### Indexing and slicing

Use square brackets to get single characters. **Counting starts at 0!**

```
 H  e  l  l  o     W  o  r  l  d
 0  1  2  3  4  5  6  7  8  9  10
-11 ...                     -2 -1
```

A **slice** takes a part of the string: `s[start:end:step]`
- `start` is **included**, `end` is **excluded**
- leave a value out to use the default (start = beginning, end = end, step = 1)

In [ ]:
hello_world = "Hello World"

print(hello_world[0])      # first character
print(hello_world[-1])     # last character
print(hello_world[0:4])    # Hell
print(hello_world[6:])     # World
print(hello_world[::2])    # every second character
print(hello_world[::-1])   # reversed!
print(hello_world[:2:2])    #start, step,stop

H
d
Hell
World
HloWrd
dlroW olleH
H


Now the string methods from the table:

In [13]:
string = "I am a string"
len(string)

13

In [14]:
splitted_string = string.split("a")
splitted_string

['I ', 'm ', ' string']

In [15]:
"_hallo_".join(splitted_string)

'I _hallo_m _hallo_ string'

In [16]:
print(string.find("am"))
print(string.find("z"))   # not found -> -1

2
-1


In [ ]:
print(string.replace("string", "Python programmer"))
print(string.upper())
print("   lots of space   ".strip())

I am a Python programmer
I AM A STRING
i am a string
lots of space


⚠️ Strings are **immutable**: methods like `replace` and `upper` return a *new* string. The original stays unchanged:

In [18]:
print(string)

I am a string


## 5. f-strings

f-strings make it easy to put variables into text. Put an `f` before the quotes and the variables in `{curly brackets}`:

In [19]:
name = "Tim"
age = 23

print(f"Hi, my name is {name} and I'm {age} years old!")

Hi, my name is Tim and I'm 23 years old!


The brackets can contain any Python expression, and you can format numbers:

In [ ]:
print(f"In two years I'll be {age + 2}!")

price = 3.14159
print(f"Rounded to 2 decimals: {price:.2f}")

In two years I'll be 25!
Rounded to 2 decimals: 3.142


## 6. Numbers

- **Integers** (`int`): whole numbers like `5`, `-12`, `0`. In Python they can be as large as you want.
- **Floats** (`float`): numbers with a decimal point like `1.45` or `-0.5`. They have limited precision.

In [21]:
an_integer = 5
a_float = 1.45

print(type(an_integer))
print(type(a_float))

<class 'int'>
<class 'float'>


### Conversion

`int()`, `float()` and `str()` convert between types.
⚠️ `int()` **cuts off** the decimal part, it does **not** round!

In [22]:
print(int(1.45))
print(int(9.99))     # 9, not 10!
print(round(9.99))   # use round() for rounding
print(float(2))

1
9
10
2.0


Strings with digits can become numbers (and the other way around):

In [23]:
print(float("1.234"))
print(int("444"))
print(str(12.345))

1.234
444
12.345


But a string with a decimal point can't go directly into `int()`. Read the error message!

In [86]:
int("1.234")

ValueError: invalid literal for int() with base 10: '1.234'

### Math

| Syntax | Meaning | Example |
|---|---|---|
| `x + y` | addition | `7 + 2` → `9` |
| `x - y` | subtraction | `7 - 2` → `5` |
| `x * y` | multiplication | `7 * 2` → `14` |
| `x / y` | division (always a float) | `7 / 2` → `3.5` |
| `x // y` | floor division | `7 // 2` → `3` |
| `x % y` | remainder (modulo) | `7 % 2` → `1` |
| `x ** y` | power | `7 ** 2` → `49` |
| `abs(x)` | absolute value | `abs(-7)` → `7` |
| `round(x, n)` | round to n decimals | `round(3.14159, 2)` → `3.14` |
| `divmod(x, y)` | `(x // y, x % y)` | `divmod(7, 2)` → `(3, 1)` |

⚠️ `x ^ y` is **not** a power in Python, it's bitwise XOR. Use `**`.

In [25]:
print(7 / 2)
print(7 // 2)
print(7 % 2)
print(2 ** 10)

3.5
3
1
1024


Python follows the usual order of operations. Use parentheses to change it:

In [26]:
print(1 + 2 * 3)
print((1 + 2) * 3)

7
9


💡 `%` is very handy, e.g. to check if a number is even: `number % 2 == 0`.

## 7. Lists, tuples and sets

### Lists
A **list** stores a sequence of items. Items can have different types, and the list can be changed (it is **mutable**).

**Syntax:** `my_list = [item0, item1, item2]`

In [27]:
list_of_arbitrary_items = [1, 2, 3.5, 4, 4, "I am a list item (*-*)"]
list_of_arbitrary_items

[1, 2, 3.5, 4, 4, 'I am a list item (*-*)']

Access items by **index** (starting at 0). Negative indices count from the end. An index that's too large raises an `IndexError`:

In [28]:
print(list_of_arbitrary_items[0])    # first item
print(list_of_arbitrary_items[3])    # fourth item
print(list_of_arbitrary_items[-1])   # last item
print(list_of_arbitrary_items[6])    # error: the list only has 6 items (index 0-5)

1
4
I am a list item (*-*)


IndexError: list index out of range

Slicing works exactly like with strings:

In [29]:
print(list_of_arbitrary_items[1:4])
print(list_of_arbitrary_items[::-1])

[2, 3.5, 4]
['I am a list item (*-*)', 4, 4, 3.5, 2, 1]


Changing a list:
- replace an item: `my_list[index] = new_value`
- add at the end: `my_list.append(new_value)`
- add at a position: `my_list.insert(index, new_value)`
- remove an item by value: `my_list.remove(value)`
- remove (and return) an item by index: `my_list.pop(index)`

In [30]:
numbers = [10, 20, 30]

numbers[1] = 99           # replace
print(numbers)

numbers.append(40)        # add at the end
print(numbers)

numbers.insert(0, 5)      # add at the beginning
print(numbers)

numbers.remove(99)        # remove by value
print(numbers)

last = numbers.pop()      # remove the last item
print(last, numbers)

[10, 99, 30]
[10, 99, 30, 40]
[5, 10, 99, 30, 40]
[5, 10, 30, 40]
40 [5, 10, 30]


`append` adds exactly **one** item. To add several, use `+=` or `.extend()`:

In [31]:
numbers = [1, 2]
numbers += [3, 4]           # or: numbers.extend([3, 4])
print(numbers)

numbers.append([5, 6])      # careful: this adds ONE item, a nested list!
print(numbers)

[1, 2, 3, 4]
[1, 2, 3, 4, [5, 6]]


Useful list tools: `len`, `index`, `count`, `sorted`, `sum`, `min`, `max`, and `in`:

In [32]:
integers = [10, 20, 30, 40, 25, 20, 100]

print(len(integers))
print(integers.index(20))   # index of the FIRST 20
print(integers.count(20))   # how many 20s?
print(sorted(integers))
print(sorted(integers, reverse=True))
print(sum(integers), min(integers), max(integers))
print(30 in integers)

7
1
2
[10, 20, 20, 25, 30, 40, 100]
[100, 40, 30, 25, 20, 20, 10]
245 10 100
True


`index` raises a `ValueError` if the value isn't in the list:

In [33]:
integers.index(420)

ValueError: 420 is not in list

### Tuples
A **tuple** looks like a list but **cannot be changed** after creation (it is **immutable**).
Use tuples for data that belongs together and shouldn't change, e.g. coordinates.

**Syntax:** `my_tuple = (item0, item1, item2)`

In [34]:
point = (3, 4)
newspapers = ("Die Presse", "Der Standard", "Kleine Zeitung")

one_element_tuple = ("singleton",)   # the comma is required!
empty_tuple = ()

print(newspapers[0])
print(type(one_element_tuple), type(("not a tuple")))

Die Presse
<class 'tuple'> <class 'str'>


Changing a tuple fails:

In [35]:
newspapers[2] = "Falter"

TypeError: 'tuple' object does not support item assignment

💡 Tuples can be **unpacked** into variables:

In [36]:
x, y = point
print(f"x = {x}, y = {y}")

x = 3, y = 4


### Sets
A **set** is an unordered collection **without duplicates**. Sets have no index.

**Syntax:** `my_set = {item0, item1, item2}`

In [88]:
a = [1, 2, 3, 4, 4, 5]
print(set(a))          # duplicates are removed

backpack = {"notebook", "phone", "key", "gum", "pen"}
print(backpack)

{1, 2, 3, 4, 5}
{'phone', 'pen', 'key', 'gum', 'notebook'}


⚠️ `{}` creates an empty **dictionary**, not an empty set. Use `set()`:

In [38]:
empty_set = set()
not_a_set = {}
print(type(empty_set), type(not_a_set))

<class 'set'> <class 'dict'>


Add with `.add()`, remove with `.remove()`:

In [39]:
backpack.add("phone charger")
backpack.remove("gum")
print(backpack)

{'phone', 'pen', 'key', 'notebook', 'phone charger'}


Running `backpack.remove("gum")` again raises a `KeyError`, because gum is already gone. `.discard()` removes without an error:

In [40]:
backpack.discard("gum")   # no error
print(backpack)

{'phone', 'pen', 'key', 'notebook', 'phone charger'}


Set elements must be **immutable** (hashable). So a list can't be inside a set:

In [41]:
{1, 2, 3, [1, 2, 3]}

TypeError: unhashable type: 'list'

**Hashable?** Python stores set elements (and dictionary keys) using a *hash*, a number computed from the value.
That number must never change, so only immutable things (numbers, strings, tuples) can be hashed.

Careful with `set()` on a single string: it splits it into letters!

In [42]:
print(set("singleton"))
print({"singleton"})

{'s', 't', 'n', 'o', 'e', 'g', 'i', 'l'}
{'singleton'}


Set operations:

In [43]:
backpack = {"notebook", "phone", "key", "gum", "wallet"}
other_bag = {"key", "sandwich", "bottle", "phone"}

print("In both (&):      ", backpack & other_bag)
print("In either (|):    ", backpack | other_bag)
print("In only one (^):  ", backpack ^ other_bag)
print("Only in first (-):", backpack - other_bag)

In both (&):       {'key', 'phone'}
In either (|):     {'phone', 'gum', 'wallet', 'bottle', 'sandwich', 'key', 'notebook'}
In only one (^):   {'bottle', 'sandwich', 'gum', 'notebook', 'wallet'}
Only in first (-): {'gum', 'notebook', 'wallet'}


## 8. Dictionaries

A **dictionary** stores **key → value** pairs. You look up values by their key, not by position.

- **Keys** must be hashable (strings, numbers, tuples) and unique.
- **Values** can be anything, even lists or other dictionaries.

In [44]:
contact_info = {
    "name": "Anna",
    "family_name": ["Smith", "Williams"],
    "phone": 12345678,
    "city": "Graz",
}
contact_info

{'name': 'Anna',
 'family_name': ['Smith', 'Williams'],
 'phone': 12345678,
 'city': 'Graz'}

Access a value with its key. A missing key raises a `KeyError`. `.get()` returns `None` (or a default) instead:

In [45]:
print(contact_info["name"])
print(contact_info.get("email"))
print(contact_info.get("email", "no email"))

Anna
None
no email


In [46]:
contact_info["email"]

KeyError: 'email'

Add, change and remove entries:

In [47]:
contact_info["mail"] = "anna@example.com"     # add
contact_info["phone"] = 98765                  # overwrite
contact_info.update({"country": "Austria", "street": "Inffeldgasse"})  # add several
del contact_info["family_name"]                # remove
print(contact_info)

{'name': 'Anna', 'phone': 98765, 'city': 'Graz', 'mail': 'anna@example.com', 'country': 'Austria', 'street': 'Inffeldgasse'}


Get all keys, values or items:

In [48]:
print(contact_info.keys())
print(contact_info.values())
print(contact_info.items())
print("city" in contact_info)    # `in` checks the KEYS

dict_keys(['name', 'phone', 'city', 'mail', 'country', 'street'])
dict_values(['Anna', 98765, 'Graz', 'anna@example.com', 'Austria', 'Inffeldgasse'])
dict_items([('name', 'Anna'), ('phone', 98765), ('city', 'Graz'), ('mail', 'anna@example.com'), ('country', 'Austria'), ('street', 'Inffeldgasse')])
True


A list can't be a key:

In [49]:
{["name", "surname"]: "Anna Smith"}

TypeError: unhashable type: 'list'

## 9. Overview: which data structure when?

| | **List** | **Tuple** | **Set** | **Dict** |
|---|---|---|---|---|
| Example | `[1, "two", True]` | `(1, "two", True)` | `{1, "two", True}` | `{"a": 1, "b": 2}` |
| Duplicates? | yes | yes | no | keys: no, values: yes |
| Ordered? | yes | yes | no | yes |
| Changeable? | yes | no | yes | yes |
| Access by | index | index | – | key |
| Empty | `[]` | `()` | `set()` | `{}` |
| Use it for | a sequence of things | fixed data | unique items | looking things up |

Data structures can be **combined**:

In [50]:
# list of lists: a matrix
identity_matrix = [
    [1, 0, 0],
    [0, 1, 0],
    [0, 0, 1],
]
print(identity_matrix[1][1])

# list of dictionaries
students = [
    {"name": "Paul", "grade": 4},
    {"name": "Lina", "grade": 1},
]
print(students[1]["name"])

# dictionary of dictionaries
registry = {
    3739: {"name": "Kem", "age": 27},
    7021: {"name": "Marie", "age": 22},
}
print(registry[7021]["age"])

1
Lina
22


## 10. Booleans

A **boolean** is either `True` or `False` (capital first letter!).

| Operation | Result |
|---|---|
| `x and y` | `True` only if **both** are `True` |
| `x or y` | `True` if **at least one** is `True` |
| `not x` | the opposite |

| x | y | `x and y` | `x or y` |
|---|---|---|---|
| False | False | False | False |
| False | True | False | True |
| True | False | False | True |
| True | True | True | True |

In [51]:
is_sunny = True
is_weekend = False

print(is_sunny and is_weekend)
print(is_sunny or is_weekend)
print(not is_sunny)

False
True
False


### Truthiness
Other values also count as `True` or `False` in conditions. Empty things and zero are `False`, everything else is `True`:

In [52]:
print(bool(0), bool(42))
print(bool(""), bool("hi"))
print(bool([]), bool([1, 2]))

False True
False True
False True


## 11. Comparisons

Every comparison gives a boolean.

| Operator | Meaning | Example |
|---|---|---|
| `<` | less than | `5 < 4` → `False` |
| `<=` | less than or equal | `5 <= 5` → `True` |
| `>` | greater than | `5 > 4` → `True` |
| `>=` | greater than or equal | `5 >= 6` → `False` |
| `==` | equal | `5 == 4` → `False` |
| `!=` | not equal | `5 != 4` → `True` |

⚠️ `=` **assigns** a value, `==` **compares** two values.

Comparisons can be chained: `0 < x < 10`

In [53]:
x = 7
print(x == 7)
print(0 < x < 10)
print("apple" < "banana")   # strings compare alphabetically

True
True
True


### Be careful when comparing floats!

In [54]:
0.1 + 0.1 + 0.1 == 0.3

False

In [55]:
0.1 + 0.1 + 0.1

0.30000000000000004

Floats are stored with limited precision, so tiny rounding errors happen
([more details](https://docs.python.org/3/tutorial/floatingpoint.html)).
Check whether two floats are **close enough** instead:

In [56]:
import math

print(abs((0.1 + 0.1 + 0.1) - 0.3) < 1e-10)
print(math.isclose(0.1 + 0.1 + 0.1, 0.3))

True
True


## 12. Conditions: `if`, `elif`, `else`

```python
if condition:
    # runs if condition is True
elif other_condition:
    # runs if the first was False and this one is True
else:
    # runs if none of the above were True
```

- Conditions are checked **from top to bottom**; only the **first** matching block runs.
- The line ends with a colon `:`.
- The code inside is **indented by 4 spaces**.

Change `a_number` and run the cells again!

In [57]:
a_number = 10

if a_number == 10:
    print("It is 10")

It is 10


In [58]:
a_number = 11

if a_number == 10:
    print("It is 10")
else:
    print("It is not 10")

It is not 10


In [59]:
a_number = 5

if a_number == 10:
    print("It is 10")
elif a_number == 5:
    print("It is not 10 but 5")
else:
    print("It is neither 10 nor 5")

It is not 10 but 5


Combine conditions with `and` / `or`:

In [60]:
temperature = 22
is_raining = False

if temperature > 18 and not is_raining:
    print("Let's go to the Schlossberg!")
else:
    print("Let's stay in the library.")

Let's go to the Schlossberg!


Use `in` / `not in` to check if something is in a list, tuple, set, string or dict:

In [61]:
number = 3
small_primes = [2, 3, 5, 7]

if number in small_primes:
    print(f"{number} is a small prime number!")

3 is a small prime number!


In [62]:
language_to_check = "french"
languages_i_speak = ["german", "english", "spanish"]

if language_to_check not in languages_i_speak:
    print(f"I've yet to learn {language_to_check}")
else:
    print(f"Yay, I know {language_to_check}")

I've yet to learn french


## 13. Common mistakes and how to fix them

**Read the error message!** The last line tells you the error type, and the `^` often points to the problem.

### Missing comma → `SyntaxError`

In [63]:
some_data = (4, 8 15, 16, 23, 42)

SyntaxError: invalid syntax. Perhaps you forgot a comma? (967563114.py, line 1)

In [64]:
some_data = (4, 8, 15, 16, 23, 42)   # fixed

### Missing colon → `SyntaxError`

In [65]:
condition = 5 > 4
if condition
    print("The condition is True")

SyntaxError: expected ':' (1364535004.py, line 2)

In [66]:
condition = 5 > 4
if condition:
    print("The condition is True")   # fixed

The condition is True


### Wrong indentation → `IndentationError`

In [67]:
some_var = 17

if some_var > 10:
print("some_var is bigger than 10.")

IndentationError: expected an indented block after 'if' statement on line 3 (166920001.py, line 4)

In [68]:
some_var = 17

if some_var > 10:
    print("some_var is bigger than 10.")   # fixed: 4 spaces

some_var is bigger than 10.


### `=` instead of `==`

In [69]:
x = 5
if x = 5:
    print("five")

SyntaxError: invalid syntax. Maybe you meant '==' or ':=' instead of '='? (4263713619.py, line 2)

### Mixing strings and numbers → `TypeError`

In [70]:
age = 23
print("I am " + age + " years old")

TypeError: can only concatenate str (not "int") to str

In [71]:
print("I am " + str(age) + " years old")
print(f"I am {age} years old")   # nicer

I am 23 years old
I am 23 years old


### The order of conditions matters
This code should print the generation for a birth year. Both versions run, but only one is correct:

In [72]:
birth_year = 1960

if birth_year <= 1964:
    print("Baby Boomer")
elif birth_year <= 1980:
    print("Generation X")
elif birth_year <= 1996:
    print("Generation Y")
elif birth_year <= 2012:
    print("Generation Z")
else:
    print("Generation Alpha")

Baby Boomer


In [73]:
birth_year = 1960

# WRONG order: the first condition is True for almost every year
if birth_year <= 2012:
    print("Generation Z")
elif birth_year <= 1996:
    print("Generation Y")
elif birth_year <= 1980:
    print("Generation X")
elif birth_year <= 1964:
    print("Baby Boomer")

Generation Z


## 14. Best practices

### Long lists: one item per line
```python
# Don't
countries = ["Afghanistan", "Albania", "Algeria", "Andorra", "Angola", "Argentina", "Armenia", "Austria"]

# Do
countries = [
    "Afghanistan",
    "Albania",
    "Algeria",
    "Austria",
]
```

### Adding one dictionary entry
```python
# Don't
contact_info.update({"mail": "anna@example.com"})

# Do
contact_info["mail"] = "anna@example.com"
```

### Avoid deep nesting
```python
# Don't
if arg1 >= 10:
    if arg2 < 0:
        if arg3 == 2:
            do_something()
        else:
            print(f"Argument 3 must be 2 but is {arg3}")
    else:
        print(f"Argument 2 must be smaller than 0 but is {arg2}")
else:
    print(f"Argument 1 must be at least 10 but is {arg1}")

# Do
if arg1 < 10:
    print(f"Argument 1 must be at least 10 but is {arg1}")
elif arg2 >= 0:
    print(f"Argument 2 must be smaller than 0 but is {arg2}")
elif arg3 != 2:
    print(f"Argument 3 must be 2 but is {arg3}")
else:
    do_something()
```

### Meaningful names
Use names that say what a variable contains: `student_count` instead of `x`, `prices` instead of `l`.
Python style: `lowercase_with_underscores`.

## 15. Exercises 📝

Try these on your own first. Ask if you get stuck!

**Exercise 1 – Strings**
Given `sentence = "Python is fun at TU Graz"`:
1. Print the number of characters.
2. Print the sentence in upper case.
3. Print only the word `"Python"` using slicing.
4. Print the sentence reversed.
5. Split it into a list of words and print how many words there are.

In [74]:
sentence = "Python is fun at TU Graz"

# your code here

**Exercise 2 – Numbers and f-strings**
A coffee costs `3.40` euros. You buy `7` coffees and pay with a 50 euro note.
Print: `7 coffees cost 23.80 €. Change: 26.20 €` (use an f-string with 2 decimals).

In [75]:
price = 3.40
count = 7
paid = 50

# your code here

**Exercise 3 – Lists**
1. Create a list `grades = [2, 1, 4, 3, 1, 5, 2]`.
2. Add the grade `1` at the end.
3. Print the best (smallest) and worst (largest) grade.
4. Print the average grade, rounded to 2 decimals.
5. Print how many students got a `1`.

In [76]:
# your code here

**Exercise 4 – Sets**
Two students list the courses they take:
```python
anna = {"Math", "Python", "Physics", "English"}
ben = {"Python", "Chemistry", "English", "History"}
```
Print the courses they **share**, all courses **together**, and the courses only **Anna** takes.

In [77]:
anna = {"Math", "Python", "Physics", "English"}
ben = {"Python", "Chemistry", "English", "History"}

# your code here

**Exercise 5 – Dictionaries**
1. Create a dictionary `student` with the keys `"name"`, `"age"` and `"courses"` (a list).
2. Add a new course to the list.
3. Add a key `"city"` with the value `"Graz"`.
4. Print: `<name> is <age> years old and takes <number> courses.`

In [78]:
# your code here

**Exercise 6 – Conditions**
Write code that prints the Austrian grade for a score (0–100):

| Score | Grade |
|---|---|
| 88–100 | 1 (Sehr gut) |
| 75–87 | 2 (Gut) |
| 63–74 | 3 (Befriedigend) |
| 50–62 | 4 (Genügend) |
| below 50 | 5 (Nicht genügend) |

Also print an error message if the score is below 0 or above 100. Test with several values.

In [79]:
score = 81

# your code here

**Bonus – Even or odd?**
Given a number, print whether it is even or odd **and** whether it is divisible by 3. (Hint: `%`)

In [80]:
number = 21

# your code here

---
## Further resources
- [Official Python tutorial](https://docs.python.org/3/tutorial/)
- [Python docs – built-in types](https://docs.python.org/3/library/stdtypes.html)
- [PEP 8 – style guide](https://peps.python.org/pep-0008/)

Questions? Ask during the tutorial or in the course forum. Happy coding! 🐍